src/solver/optimisation/opt_gestion_second_membre_cas_lineaire.cpp

```diff
@@ -138,240 +166,758 @@ void OPT_InitialiserLeSecondMembreDuProblemeLineaire(PROBLEME_HEBDO* problemeHeb
+            namer.UpdateTimeStep(problemeHebdo->weekInTheYear);
+            namer.MaxPumping(problemeHebdo->ProblemeAResoudre->NombreDeContraintes);
+
+            const double rhs = problemeHebdo->CaracteristiquesHydrauliques[pays]
+                                 .MaxEnergiePompageParIntervalleOptimise[NumeroDeLIntervalle];
+            builder.lessThan(rhs).build();
         }
     }
+};
 
-    for (int pdtJour = 0, pdtHebdo = PremierPdtDeLIntervalle; pdtHebdo < DernierPdtDeLIntervalle;
-         pdtHebdo++, pdtJour++)
+struct AreaHydroLevel : public Constraint
+{
+    using Constraint::Constraint;
+    void add(int pays, int pdt, int pdtHebdo)
     {
-        const CORRESPONDANCES_DES_CONTRAINTES& CorrespondanceCntNativesCntOptim
-          = problemeHebdo->CorrespondanceCntNativesCntOptim[pdtJour];
-
-        for (int pays = 0; pays < problemeHebdo->NombreDePays; pays++)
+        double rhs
+          = problemeHebdo->CaracteristiquesHydrauliques[pays].ApportNaturelHoraire[pdtHebdo];
+        if (pdtHebdo == 0)
         {
-            if (!problemeHebdo->CaracteristiquesHydrauliques[pays].SuiviNiveauHoraire)
-                continue;
+            rhs += problemeHebdo->CaracteristiquesHydrauliques[pays].NiveauInitialReservoir;
+        }
+        builder.updateHourWithinWeek(pdt).include(Variable::HydroLevel(pays), 1.0);
+        if (pdt > 0)
+        {
+            builder.updateHourWithinWeek(pdt - 1).include(Variable::HydroLevel(pays), -1.0);
+        }
+        builder.updateHourWithinWeek(pdt)
+          .include(Variable::HydProd(pays), 1.0)
+          .include(Variable::Pumping(pays),
+                   -problemeHebdo->CaracteristiquesHydrauliques[pays].PumpingRatio)
+          .include(Variable::Overflow(pays), 1.)
+          .equalTo(rhs)
+          .build();
+        ConstraintNamer namer(problemeHebdo->ProblemeAResoudre->NomDesContraintes,
+                              problemeHebdo->NamedProblems);
+
+        namer.UpdateArea(problemeHebdo->NomsDesPays[pays]);
+        namer.UpdateTimeStep(problemeHebdo->weekInTheYear * 168 + pdt);
+        namer.AreaHydroLevel(problemeHebdo->ProblemeAResoudre->NombreDeContraintes);
+    }
+};
 
-            int cnt = CorrespondanceCntNativesCntOptim.NumeroDeContrainteDesNiveauxPays[pays];
-            if (cnt >= 0)
-            {
-                SecondMembre[cnt] = problemeHebdo->CaracteristiquesHydrauliques[pays]
-                                      .ApportNaturelHoraire[pdtHebdo];
-                if (pdtHebdo == 0)
-                {
-                    SecondMembre[cnt]
-                      += problemeHebdo->CaracteristiquesHydrauliques[pays].NiveauInitialReservoir;
-                }
-                AdresseOuPlacerLaValeurDesCoutsMarginaux[cnt] = nullptr;
-            }
+struct FinalStockEquivalent : public Constraint
+{
+    using Constraint::Constraint;
+    void add(int pays)
+    {
+        const auto& pdt = problemeHebdo->NombreDePasDeTempsPourUneOptimisation - 1;
```

Comment:
```suggestion
        const auto pdt = problemeHebdo->NombreDePasDeTempsPourUneOptimisation - 1;
```
Not really an lvalue because of the "-1".

Reply:
definitely